File: apps/web/lib/api/attachments.ts

import { getFetchClient } from "@/lib/fetch";
import { getApiProxyUrl } from "@/lib/url";
import type { Attachment, AttachmentWithMaybeTagsAndSource, AttachmentWithTags, UpdateAttachment } from "@/types/attachments";
import type { InboxItem } from "@/types/inbox";

const fetch = <T,>(...args: Parameters<ReturnType<typeof getFetchClient<T>>>) => getFetchClient<T>()(...args)

export async function updateAttachment(attachmentID: string, body: UpdateAttachment): Promise<AttachmentWithMaybeTagsAndSource | null> {
  try {
    const { data, error } = await fetch<AttachmentWithMaybeTagsAndSource | null>({
      url: getApiProxyUrl(`/v1/my/attachments/${attachmentID}`),
      method: "PATCH",
      body: body
    })

    if (error) {
      throw error;
    }

    return data;

  } catch (error) {
    console.error(error)
    return null
  }
}

export async function publishAttachment(attachmentID: string) {
  return updateAttachment(attachmentID, { public: true })
}

export async function getInboxAttachments(tags?: string[]): Promise<InboxItem[] | null> {
  try {
    const url = new URL(getApiProxyUrl("/v1/my/attachments/inbox"))
    if (tags?.length) url.searchParams.set("tags", tags.join(","))

    const { data, error } = await fetch<InboxItem[]>({
      url: url.toString()
    })

    if (error) {
      throw error;
    }

    return data;

  } catch (error) {
    console.error(error)
    return null
  }
}

export async function getAttachment(attachmentID: string): Promise<AttachmentWithTags | null> {
  try {
    const { data, error } = await fetch<AttachmentWithTags>({
      url: getApiProxyUrl(`/v1/my/attachments/${attachmentID}`)
    })

    if (error) {
      throw error;
    }

    return data;

  } catch (error) {
    console.error(error)
    return null
  }
}

export async function getPublicAttachment(attachmentID: string): Promise<AttachmentWithTags | null> {
  try {
    const { data, error } = await fetch<AttachmentWithTags>({
      url: getApiProxyUrl(`/v1/attachments/${attachmentID}`)
    })

    if (error) throw error
    return data
  } catch (error) {
    console.error(error)
    return null
  }
}

export async function moveAttachmentToCollection(attachmentID: string, collectionID: string): Promise<unknown> {
  try {
    const { data, error } = await fetch<unknown>({
      url: getApiProxyUrl(`/v1/my/collections/${collectionID}/attachments?attachmentID=${attachmentID}`),
      method: "POST",
    })

    if (error) {
      throw error;
    }

    return data;

  } catch (error) {
    console.error(error)
    return null;
  }
}

export async function removeAttachmentFromCollection(attachmentID: string, collectionID: string): Promise<unknown> {
  try {
    const { data, error } = await fetch<unknown>({
      url: getApiProxyUrl(`/v1/my/collections/${collectionID}/attachments?attachmentID=${attachmentID}`),
      method: "DELETE",
    })

    if (error) {
      throw error;
    }

    return data;

  } catch (error) {
    console.error(error)
    return null;
  }
}

export async function getCollectionAttachments(collectionID: string, tags?: string[]): Promise<AttachmentWithTags[] | null> {
  try {
    const url = new URL(getApiProxyUrl(`/v1/my/collections/${collectionID}/attachments`))
    if (tags?.length) url.searchParams.set("tags", tags.join(","))

    const { data, error } = await fetch<AttachmentWithTags[]>({
      url: url.toString(),
    })

    if (error) {
      throw error;
    }

    return data;

  } catch (error) {
    console.error(error)
    return null;
  }
}

export async function getPublicCollectionAttachments(collectionID: string): Promise<AttachmentWithTags[] | null> {
  try {
    const { data, error } = await fetch<AttachmentWithTags[]>({
      url: getApiProxyUrl(`/v1/collections/${collectionID}/attachments`),
    })
    // console.log(getApiProxyUrl(`/v1/collections/${collectionID}/attachments`))

    if (error) {
      throw error;
    }

    return data;

  } catch (error) {
    console.error(error)
    return null;
  }
}

type ListQuery = {
  offset?: number;
  limit?: number;
  tags?: string[];
}
export async function getAllAttachments(query?: ListQuery): Promise<AttachmentWithTags[] | null> {
  try {
    const url = new URL(getApiProxyUrl("/v1/my/attachments"))
    if (query) {
      if (query.offset !== undefined) url.searchParams.set("offset", query.offset.toString())
      if (query.limit !== undefined) url.searchParams.set("limit", query.limit.toString())
      if (query.tags?.length) url.searchParams.set("tags", query.tags.join(","))
    }
    const { data, error } = await fetch<AttachmentWithTags[]>({
      url: url.toString(),
    })

    if (error) {
      throw error;
    }

    return data;

  } catch (error) {
    console.error(error)
    return null;
  }
}

export async function permanentlyDeleteAttachment(attachmentID: string): Promise<{ id: string } | null> {
  try {
    const { data, error } = await fetch<{ id: string } | null>({
      url: getApiProxyUrl(`/v1/my/attachments/${attachmentID}`),
      method: "DELETE",
    })

    if (error) {
      throw error;
    }

    return data;

  } catch (error) {
    console.error(error)
    return null;
  }
}

export async function moveToTrashAttachment(attachmentID: string): Promise<{ id: string } | null> {
  try {
    const { data, error } = await fetch<{ id: string } | null>({
      url: getApiProxyUrl(`/v1/my/attachments/${attachmentID}/trash`),
      method: "POST",
    })

    if (error) {
      throw error;
    }

    return data;

  } catch (error) {
    console.error(error)
    return null;
  }
}

export async function restoreAttachment(attachmentID: string): Promise<{ id: string } | null> {
  try {
    const { data, error } = await fetch<{ id: string } | null>({
      url: getApiProxyUrl(`/v1/my/attachments/${attachmentID}/untrash`),
      method: "POST",
    })

    if (error) throw error
    return data
  } catch (error) {
    console.error(error)
    return null
  }
}

export async function getTrashAttachments(): Promise<AttachmentWithMaybeTagsAndSource[] | null> {
  try {
    const { data, error } = await fetch<AttachmentWithMaybeTagsAndSource[]>({
      url: getApiProxyUrl("/v1/my/attachments/trash")
    })

    if (error) {
      throw error;
    }

    return data;

  } catch (error) {
    console.error(error)
    return null
  }
}

export async function uploadAttachment(file: Blob): Promise<Attachment | null> {
  try {
    const formData = new FormData()

    formData.append("file", file)

    const { data, error } = await fetch<Attachment | null>({
      url: getApiProxyUrl("/v1/my/attachments/new"),
      method: "POST",
      body: formData,
    });

    if (error) {
      throw error;
    }

    return data;

  } catch (error) {
    console.error(error)
    return null
  }
}

export async function inboxAttachment(id: string) {
  try {
    const { status } = await fetch<null>({
      url: getApiProxyUrl(`/v1/my/attachments/inbox?attachmentID=${id}`),
      method: "POST",
    });
    return { status };
  } catch (error) {
    console.error(error)
    return null
  }
}

File: apps/api/internal/services/attachments.go
package services

import (
	"github.com/google/uuid"
	"github.com/yz13-dev/imc/api/internal/models"
	"github.com/yz13-dev/imc/api/internal/repositories"
	"gorm.io/gorm"
)

func GetInboxAttachments(UserID string, db *gorm.DB) ([]models.InboxItem, error) {
	attachments, err := repositories.GetInboxAttachments(UserID, db)
	if err != nil {
		return nil, err
	}
	return attachments, nil
}

func PostNewAttachment(UserID string, db *gorm.DB, data models.NewAttachment) (models.Attachment, error) {
	attachment, err := repositories.PostNewAttachment(UserID, db, data)
	if err != nil {
		return models.Attachment{}, err
	}
	return attachment, nil
}

func GetAttachment(UserID string, attachmentID string, db *gorm.DB) (models.AttachmentWithTags, error) {
	attachment, err := repositories.GetAttachment(UserID, attachmentID, db)
	if err != nil {
		return models.AttachmentWithTags{}, err
	}
	return attachment, nil
}

func GetPublicAttachment(attachmentID uuid.UUID, db *gorm.DB) (*models.AttachmentWithTags, error) {
	attachment, err := repositories.GetPublicAttachment(attachmentID, db)
	if err != nil {
		return nil, err
	}
	return attachment, nil
}

func PatchAttachment(AttachmentID uuid.UUID, UserID string, data models.UpdateAttachment, db *gorm.DB) (models.Attachment, error) {
	attachment, err := repositories.PatchAttachment(AttachmentID, UserID, data, db)
	if err != nil {
		return models.Attachment{}, err
	}
	return attachment, nil
}

func PostInInbox(UserID string, db *gorm.DB, attachmentID uuid.UUID) error {
	err := repositories.PostInInbox(UserID, db, attachmentID)
	if err != nil {
		return err
	}
	return nil
}

func GetCollectionAttachments(collectionID uuid.UUID, UserID string, db *gorm.DB) ([]models.AttachmentWithTags, error) {
	attachments, err := repositories.GetCollectionAttachments(collectionID, UserID, db)
	if err != nil {
		return nil, err
	}
	return attachments, nil
}

func GetAllAttachments(UserID string, query repositories.ListQuery, db *gorm.DB) ([]models.AttachmentWithTags, error) {
	attachments, err := repositories.GetAllAttachments(UserID, query, db)
	if err != nil {
		return nil, err
	}
	return attachments, nil
}

func TrashAttachment(UserID string, attachmentID string, db *gorm.DB) error {
	err := repositories.TrashAttachment(UserID, attachmentID, db)
	if err != nil {
		return err
	}
	return nil
}

func UntrashAttachment(UserID string, attachmentID string, db *gorm.DB) error {
	err := repositories.UntrashAttachment(UserID, attachmentID, db)
	if err != nil {
		return err
	}
	return nil
}

func DeleteAttachment(UserID string, attachmentID string, db *gorm.DB) (models.Attachment, error) {
	attachment, err := repositories.DeleteAttachment(UserID, attachmentID, db)
	if err != nil {
		return models.Attachment{}, err
	}
	return attachment, nil
}

func GetTrashAttachments(UserID string, db *gorm.DB) ([]models.AttachmentWithTags, error) {
	attachments, err := repositories.GetTrashAttachments(UserID, db)
	if err != nil {
		return nil, err
	}
	return attachments, nil
}

func GetAttachmentWithCollection(ID uuid.UUID, db *gorm.DB) (*models.CollectionAttachmentWithAttachmentAndAttachment, error) {
	attachment, err := repositories.GetAttachmentWithCollection(ID, db)
	if err != nil {
		return nil, err
	}
	return attachment, nil
}
